### PR TITLE
Change implementation from Graph to matrix

### PR DESCRIPTION
**Pull request by lima-limon-inc**
- Instead of representing the Merkle tree with a graph, now I'm storing it as a matrix. This made the code x10 cleaner and easier to understand.
- Now, I'll tackle producing verification next.

### CODE REVIEW

File: src/merkle_tree.rs [hunk 1]
```diff
@@ -1,127 +1,78 @@
...
+        tree.push(new_leaves.clone());
         loop {
-	  tree.extend(latest_leaves.clone());
-	  latest_leaves = Self::add_children_leaves(latest_leaves, tree.len());
-	  if latest_leaves.len() == 1 {
-	      tree.extend(latest_leaves.clone());
+
+	  new_leaves = new_leaves
+	      .chunks(2)
+	      .map(|left_n_right|
```
Comment (dsocolobsky) [inline src/merkle_tree.rs:38]: `|(left, right)|` maybe is better to avoid indexing?

File: src/merkle_tree.rs [hunk 2]
```diff
@@ -1,127 +1,78 @@
...
+		 {
+		     if left_n_right.get(1).is_none() {
+		        // This is the case where we have an uneven
+		        // amount of nodes
+		         hash(&[left_n_right[0]])
+                             // [&left_n_right[0], &left_n_right[0]]
+
```
Comment (damiramirez) [inline src/merkle_tree.rs]: Could it be possible that you forgot to duplicate the hash? Maybe something like this: `hash(&[left_n_right[0], left_n_right[0]])`

File: src/merkle_tree.rs [hunk 3]
```diff
@@ -1,127 +1,78 @@
...
-        for leaf in tree.clone() {
-	  println!("{:?}", leaf);
+        println!("HERE");
+        println!("{:?}", tree);
+        for level in tree.clone() {
+	  // for node in level {
+	  //     print!("{:?}", node);
+	  // }
+	  println!("{}", level.len())
         }
```
Comment (damiramirez) [inline src/merkle_tree.rs:91]: You can create a function to perform printing. Or maybe implement a Display trait.

**Review by damiramirez — CHANGES_REQUESTED** [review]

File: src/merkle_tree.rs [hunk 4]
```diff
@@ -1,127 +1,78 @@
...
+
+	  new_leaves = new_leaves
+	      .chunks(2)
+	      .map(|left_n_right|
+		 {
+		     if left_n_right.get(1).is_none() {
+		        // This is the case where we have an uneven
+		        // amount of nodes
+		         hash(&[left_n_right[0]])
+                             // [&left_n_right[0], &left_n_right[0]]
```
Comment (dsocolobsky) [inline src/merkle_tree.rs]: This comment is confusing, looks like a line of code, I would delete it if you are not using it.

File: src/merkle_tree.rs (hunk 2 above)
Comment (lima-limon-inc) [reply src/merkle_tree.rs]: Yes indeed! Great catch. Thanks

File: src/merkle_tree.rs [hunk 5]
```diff
@@ -1,127 +1,78 @@
...
-
-pub fn hash(positions: [Position; 2], leaves: &[&Leaf; 2]) -> Leaf {
+pub fn hash(leaves: &[HashedData]) -> HashedData {
     let new_hash = leaves
         .iter()
         .fold(Sha3_256::new(), |mut acc, a| {
-            let value = a.hash;
+            let value = a;
             acc.update(value);
             acc
```
Comment (LeanSerra) [inline src/merkle_tree.rs]: `a` could have a better name

File: src/merkle_tree.rs [hunk 6]
```diff
@@ -1,127 +1,78 @@
...
 
-        println!("");
-        for leaf in tree.clone() {
-	  println!("{:?}", leaf);
+        println!("HERE");
+        println!("{:?}", tree);
+        for level in tree.clone() {
+	  // for node in level {
+	  //     print!("{:?}", node);
+	  // }
```
Comment (dsocolobsky) [inline src/merkle_tree.rs]: Commented code

File: src/merkle_tree.rs [hunk 7]
```diff
@@ -1,127 +1,78 @@
...
+	  println!("{}", level.len())
         }
-
-        let leaves: Vec<Leaf> = tree
-	  .into_iter()
-	  .map(|a| a.1)
-	  .collect();
+        // for leaf in tree.clone().iter().flatten().collect() {
+        // 	  println!("{:?}", leaf);
+        // }
```
Comment (dsocolobsky) [inline src/merkle_tree.rs]: Commented code

**Review by dsocolobsky — CHANGES_REQUESTED** [review]

File: src/merkle_tree.rs (hunk 3 above)
Comment (lima-limon-inc) [reply src/merkle_tree.rs:91]: Thanks for the tip! I'll implement it later on however

**Review by LeanSerra — CHANGES_REQUESTED** [review]

File: src/merkle_tree.rs (hunk 4 above)
Comment (lima-limon-inc) [reply src/merkle_tree.rs]: Fixed in latest commit

File: src/merkle_tree.rs (hunk 5 above)
Comment (lima-limon-inc) [reply src/merkle_tree.rs]: Fixed in latest commit

**Review by damiramirez — APPROVED** [review]

**Review by dsocolobsky — APPROVED** [review]

**Pull request merged**